Fusing Initializers with Graph Transforms

sunnyshu-intel opened the pull request:
### Description

Added a graph transform for mixed precision graphs when FP16 compute is unavailable. At session creation, this graph transform converts FP16 initializers (_which were changed to FP16 to FP32 cast nodes_) to FP32 initializers and fuses them with their next FP32 nodes.
 

- Behavior before this change:
"fp16 initializers -> cast_from_fp16_to_fp32 -> fp32 node/s"
 
- Behavior after this change:
"fp16 initializers converted to fp32 initializers then fused with fp32 node/s"

### Motivation and Context

This change aims to run the FP16 models without the repetitive casting of FP16 initializers to FP32 initializers, by fusing FP32 initializers with their next nodes, when FP16 compute is not available.

> For naming purposes, the newly added Graph Transforms in long form is called "Fused Initializers Graph Transforms", and in short form is called "FIGT".

### Working

Currently, the Fuse Initializers Graph Transform fuses cast nodes that casts from FP16 to FP32, back to their
next/output nodes. Below is an explanation of how this transforms works. It depends on ```InsertCastTransforms```
to produce the intermediate representation from which it fuses the initializers (which are the cast node with
zero input, one initializer, and one output) back to the next/output node. After fusion, the link/edge between such
a cast node to the next/output node will then be removed. Cast nodes will be removed as well.

```
        "Input Graph"                       "Intermediate Representation"                 "FIGT Transforms"

          --------                   --------        --------        --------                 --------
         | X_Fp16 |                 | X_Fp16 |      | W_Fp16 |      | B_Fp16 |               | X_Fp16 |
          --------                   --------        --------        --------                 --------
             |                          |               |               |                        |
             |                          |               |               |                        |
             |                          V               V               V                        V
             |                       | Cast |        | Cast |        | Cast |                 | Cast |
             |                       | Fp16 |        | Fp16 |        | Fp16 |                 | Fp16 |
             |                       |  To  |        |  To  |        |  To  |                 |  To  |
             |                       | Fp32 |        | Fp32 |        | Fp32 |                 | Fp32 |
             |                          |               |               |                        |
             |                          |               |               |                        |
             V                          V               V               V                        V
 ----------------------------       -----------------------------------------       ----------------------------
|        Conv_Fp16           |     |                                         |     |         Conv_Fp32          |
|        --W_Fp16--          | ==> |                Conv_Fp32                | ==> |         --W_Fp32--         |
|        --B_Fp16--          |     |                                         |     |         --B_Fp32--         |
 ----------------------------       -----------------------------------------       ----------------------------
             |                                          |                                        |
             |                                          |                                        |
             |                                          V                                        V
             |                                       | Cast |                                 | Cast |
             |                                       | Fp32 |                                 | Fp32 |
             |                                       |  To  |                                 |  To  |
             |                                       | Fp16 |                                 | Fp16 |
             |                                          |                                        |
             |                                          |                                        |
             V                                          V                                        V
          --------                                   --------                                 --------
         | Y_Fp16 |                                 | Y_Fp16 |                               | Y_Fp16 |
          --------                                   --------                                 --------
```

The newly added Graph Transforms perform the following actions.

* Detect Cast node/s with single FP16 initializer converting to FP32.
* Convert all such FP16 initializer/s to FP32 initializer/s.
* Fuse newly created FP32 initializer/s to relative FP32 node/s.
* Remove FP16 to FP32 Cast node/s.

This is run in a loop as follows. It excludes Level 1 and Partitioning optimizations.

```
 Level 2 --> Level 3 --> InsertCastTransforms --> FIGT
   ^                                        	   |
   |                 "LOOP"                	   |
   |                                       	   |
   -------------------------------------------------
```

### Adding FIGT as a Level-4 Graph Transform.

This will have the following benefits.

1. Ability to turn off (any/all) the Level 4 Optimizations. We can use the `disable optimizers` functionality to turn off one of such optimizations during testing, or use the `-o` switch to turn off all Level 4 optimizations while executing a model using the command line or Python scripts (or any other scripts).

2. Ability to rerun Level 2 and Level 3 optimizations remains intact after Level 4 Optimizations are applied. Adding Level 4 takes care that FIGT (or any similar optimizations) will always run after InsertCastNodes.

3. It keeps the current graph manipulations untouched and gives us more flexibility to add future optimizations like adding `Int8 to Int32` upconvert or `FP8 to FP16` upconvert under Level 4. Level 4 can, as of now, work as a placeholder for any other such upcoming Graph optimizations.

```
 Level 2 --> Level 3 --> InsertCastTransforms --> Level 4
   ^                                        	    |
   |                 "LOOP"                	    |
   |                                       	    |
   --------------------------------------------------
```
> Added a placeholder for Level-4 for graph transforms utils under orttraining. This helps resolve any exceptions that may be encountered during training sessions.

#### Re-running Level 2+ optimizations after Level 4 / FIGT

The idea behind re-running Level2+ graph transforms is that, after the fusion of initializers with their respective nodes, the nodes are now in a format that might be supported by other graph transforms that were previously skipped. Hence, some of the transformations previously unable to be applied are now valid and can be applied to create a more optimal graph for execution.

### Added a new session option "kOrtSessionOptionsGraphOptimizationsLoopLevel" to handle the graph optimization loop.

* When set to 2 or above it will loop until no more optimizations are applied at any level starting Level 2 and above.

```
  Level 2 --> Level 3 --> InsertCastTransforms --> Level 4
    ^                                                 |
    |                 "Loop"                          |
    |                                                 |
    ---------------------------------------------------
```

* When set to 1 (default) it will loop until no more optimizations are applied at Level 4 only.

```
  Level 2 --> Level 3 --> InsertCastTransforms --> Level 4
    ^                                                 |
    |        "Loop only depending on Level 4"         |
    |                                                 |
    ---------------------------------------------------
```

* When set to 0 it disables the loop.

```
  Level 2 --> Level 3 --> InsertCastTransforms --> Level 4
    ^                                                 |
    |                 "No Loop"                       |
    |                                                 |
    X                xxxxxxxxxxx                      X
```

### Documentation

We have not added any details related to Level 4 in the [Graph Optimizations in ONNX Runtime](https://onnxruntime.ai/docs/performance/model-optimizations/graph-optimizations.html) documentation.

### OLD PR

This PR is created following a thorough discussion on the [OLD PR](https://github.com/microsoft/onnxruntime/pull/24175).

tianleiwu commented:
Please follow https://github.com/microsoft/onnxruntime/blob/be20f165a93e2219718d8c4fa002e6b678077548/docs/Coding_Conventions_and_Standards.md#linting to format the code.

Please take a look at failed checks, and fix errors that caused by this PR.

sunnyshu-intel commented:
> Please follow https://github.com/microsoft/onnxruntime/blob/be20f165a93e2219718d8c4fa002e6b678077548/docs/Coding_Conventions_and_Standards.md#linting to format the code.
> 
> Please take a look at failed checks, and fix errors that caused by this PR.

Hi, we are trying to solve the build errors from our side. I will take a few days to fix errors and upload the fixes.

sunnyshu-intel commented:
> Please follow https://github.com/microsoft/onnxruntime/blob/be20f165a93e2219718d8c4fa002e6b678077548/docs/Coding_Conventions_and_Standards.md#linting to format the code.
> 
> Please take a look at failed checks, and fix errors that caused by this PR.

Hi, I have pushed updates to address the build errors in this PR. Could you please re-run the workflow checks? Please note that I uploaded some intermediate updates to this branch, which may have reset certain checks. As a result, I may have missed some build errors. I will make sure to avoid uploading intermediate updates in the future. Once the workflow checks are complete, I will promptly address any remaining errors that may have been missed due to the reset. Thanks.

sunnyshu-intel commented:
> Please follow https://github.com/microsoft/onnxruntime/blob/be20f165a93e2219718d8c4fa002e6b678077548/docs/Coding_Conventions_and_Standards.md#linting to format the code.
> 
> Please take a look at failed checks, and fix errors that caused by this PR.

Hi, I have pushed updates to address the build errors in this PR. Could you please re-run the workflow checks? Please note that I uploaded some intermediate updates to this branch, which reset the checks. As a result, I may have missed some build errors. I will make sure to avoid uploading intermediate updates in the future. Once the workflow checks are complete, I will promptly address any remaining errors that may have been missed due to the reset. Thanks.

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline

sunnyshu-intel commented:
> The following two tests failed in [Windows CPU CI Pipeline](https://github.com/microsoft/onnxruntime/actions/runs/15196682084/job/42763022709?pr=24726): TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_1 TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_2

Yes am currently looking into it.

sunnyshu-intel commented:
> * Linux CI / Build Linux x64 Release with training / build_test_pipeline (pull_request)

Hi there are still some pending checks. Also for
[Linux CI / Build Linux x64 Release with training / build_test_pipeline (pull_request)](https://github.com/microsoft/onnxruntime/actions/runs/15196682131/job/42763023332?pr=24726) and [Linux DNNL CI / Build Linux x64 Release (DNNL EP) / build_test_pipeline (pull_request)](https://github.com/microsoft/onnxruntime/actions/runs/15196682073/job/42763022404?pr=24726) the failure doesn't seem to be affected by the changes in this PR. Please let me know if it is fine for me to upload the updated code.

sunnyshu-intel commented:
> The following two tests failed in [Windows CPU CI Pipeline](https://github.com/microsoft/onnxruntime/actions/runs/15196682084/job/42763022709?pr=24726): TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_1 TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_2

Hi, I pushed a fix; it should solve all the build errors. Please re-run the workflow checks. Before pushing the fix, some workflow checks were failing, and some were pending. Below is the summary.

### Failed
- Linux CI / Build Linux x64 Release with training / build_test_pipeline (pull_request)
- Linux DNNL CI / Build Linux x64 Release (DNNL EP) / build_test_pipeline (pull_request)
- Web CI Pipeline / wasm_Debug / build-wasm (pull_request)
- Windows CPU CI Pipeline / build_x86_release (pull_request)

### Pending
- license/cla
- web_Debug / build_onnxruntime_web

sunnyshu-intel commented:
> > The following two tests failed in [Windows CPU CI Pipeline](https://github.com/microsoft/onnxruntime/actions/runs/15196682084/job/42763022709?pr=24726): TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_1 TransformerTest.FuseFp16InitializersWithFp32Node_with_graph_optimizations_loop_level_set_to_2
> 
> Hi, I pushed a fix; it should solve all the build errors. Please re-run the workflow checks. Before pushing the fix, some workflow checks were failing, and some were pending. Below is the summary.
> 
> ### Failed
> * Linux CI / Build Linux x64 Release with training / build_test_pipeline (pull_request)
> * Linux DNNL CI / Build Linux x64 Release (DNNL EP) / build_test_pipeline (pull_request)
> * Web CI Pipeline / wasm_Debug / build-wasm (pull_request)
> * Windows CPU CI Pipeline / build_x86_release (pull_request)
> 
> ### Pending
> * license/cla
> * web_Debug / build_onnxruntime_web

FYI, adding to the above conversation. I am not sure why the `license/cla` check was still pending last time. It finishes and passes at first whenever I push updates. After pushing the fix, it did pass at first, as seen from the figure below.

![image](https://github.com/user-attachments/assets/7d3b250f-57f9-4ae9-996d-2898091a6db1)

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline

sunnyshu-intel commented:
> For Linux CI, seems need to update the training code to handle level 4:
> 
> ```
> C++ exception with description "/onnxruntime_src/orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:220 Unsupported level 4
> ```
> 
> For DNNL, it is not sure about the root cause. A walkaround is to skip level 4 and loop when compiling DNNL.

Hi, a workaround as follows can be implemented, which will resolve the `Unsupported level` exception. This is just adding Level 4 as a placeholder for any kind of `Unsupported Datatype Optimizations` in the training graph transformer utils. I am not adding the current `Fuse Initializers Graph Transforms` to the level 4 for training, as I am not sure how it will affect the training side of ORT, and the focus is on performance gains with `inference` for this PR.

```diff
--- a/orttraining/orttraining/core/optimizer/graph_transformer_utils.cc
+++ b/orttraining/orttraining/core/optimizer/graph_transformer_utils.cc
@@ -216,6 +216,9 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
     case TransformerLevel::Level3: {
     } break;
 
+    case TransformerLevel::Level4: {
+    } break;
+
     default:
       ORT_ENFORCE(false, "Unsupported level " + std::to_string(static_cast<uint32_t>(level)));
       break;
@@ -295,6 +298,14 @@ InlinedVector<std::unique_ptr<GraphTransformer>> GenerateTransformers(
       }
     } break;
 
+    case TransformerLevel::Level4: {
+      // NOTE: Placeholder for adding level 4 transformers to handle unsupported datatype optimizations.
+      // For inference, FuseInitializersTransformer is used to fuse FP16 initializers with FP32 nodes.
+      // For training, if it is necessary to fuse FP16 initializers to FP32 nodes (e.g., to support platforms without FP16),
+      // add FuseInitializersTransformer to the training graph transformers list.
+      // For reference, see FuseFp16InitializerToFp32NodeTransformer in onnxruntime/core/optimizer/graph_transformer_utils.cc.
+    } break;
+
     default:
       ORT_ENFORCE(false, "Unsupported level " + std::to_string(static_cast<uint32_t>(level)));
       break;
```

sunnyshu-intel commented on onnxruntime/core/optimizer/fuse_initializers_transformer.cc:
```diff
@@ -0,0 +1,172 @@
+
+  // Node must have only one output edge
+  if (!(1 == node.GetOutputEdgesCount())) return false;
+
```
Hi, the checks in the code are done by two functions. These function in combination makes sure we are looking at `Cast` nodes, which are inserted using `InsertCastTrasforms` when fp16 (in this case) compute is not supported on any platforms. These `Cast` nodes have a single fp16 initializer, which is converted to fp32 at each inference.

1. IsCastNodeWithConstraints
    - Node must be a Cast node
    - Node must have no input edges
    - Node must have only one output edge
2. IsNodeValidForFusion
    - Node must have an initialized tensor
    - Initialzed tensor must be of the fp16 type
    - Node output must be of the fp32 type

I can add one more check to the `IsNodeValidForFusion` function as follows to make sure we are targeting input nodes with only one initializer. 

```diff
--- a/onnxruntime/core/optimizer/fuse_initializers_transformer.cc
+++ b/onnxruntime/core/optimizer/fuse_initializers_transformer.cc
@@ -50,6 +50,9 @@ static bool IsNodeValidForFusion(const Graph& graph,
                                  const Node& node,
                                  const onnxruntime::MLDataType tensor_type,
                                  const onnxruntime::MLDataType output_type) {
+  // Node must have only one input definition
+  if (!(1 == node.InputDefs().size())) return false;
+
   // Node must have initialized tensor
   if (!(graph.IsInitializedTensor(node.InputDefs()[0]->Name()))) return false;
```

To the best of my knowledge, due to the above checks, we will never encounter a **graph output node**, as the nodes that are fused with this PR have no input edges.  However, if it is required to add such a check, I can add it as follows.

```diff
--- a/onnxruntime/core/optimizer/fuse_initializers_transformer.cc
+++ b/onnxruntime/core/optimizer/fuse_initializers_transformer.cc
@@ -145,8 +145,18 @@ Status FuseInitializersTransformer::ApplyImpl(Graph& graph, bool& modified, int
     // Get Node
     auto node = graph.GetNode(node_index);
 
+    // Skip if node is nullptr
+    if (!node) {
+      continue;
+    }
+
+    // Skip if node produces graph output
+    if(graph_viewer.NodeProducesGraphOutput(*node)) {
+      continue;
+    }
+
     // Check if the current node is a Cast node with all constraints and valid for fusion
-    if (node && IsCastNodeWithConstraints(*node) && IsNodeValidForFusion(graph, *node, init_type_, cvt_type_)) {
+    if (IsCastNodeWithConstraints(*node) && IsNodeValidForFusion(graph, *node, init_type_, cvt_type_)) {
       // Add node to the set of nodes to be fused and removed
       nodes_to_be_fused_and_removed_from_graph.insert(node_index);
     }
```

Please let me know how I should proceed. Thanks

tianleiwu replied:
Please consider the following test case:
```
Fp16_Initializer --> Cast (to=float) --> Identity --> graph_output_2
                                 |  
                                V
                       graph_output_1
--------------------------------------------------
Expected result after fusion:

Fp16_Initializer --> Cast (to=float)
                              |  
                              V
                       graph_output_1

Fp32_Intializer (with new name) --> Identity --> graph_output_2
```

If the output is a graph output, you will need keep the Cast node after fusion.

tianleiwu replied:
The definition of EdgeEnd is node to node, so the link of node to graph output is not counted as output edge. You can add such a test case to trace.

tianleiwu replied:
> Node must have one input def (adding it to checks)

This is not needed as I mentioned. Such constraint is from onnx spec. An invalid graph (like Cast has no input or more than one input) will fail early (during onnx model loading).

sunnyshu-intel replied:
I have added the required changes.

tianleiwu commented on onnxruntime/core/optimizer/fuse_initializers_transformer.cc:
```diff
@@ -0,0 +1,172 @@
+                                    const onnxruntime::MLDataType next_node_arg_type,
+                                    onnxruntime::concurrency::ThreadPool* thread_pool) {
+  // Get next node
+  Node& next_node = *graph.GetNode(node.OutputNodesBegin()->Index());
```
There might not be next node: graph.GetNode(node.OutputNodesBegin()->Index()) could be nullptr.

sunnyshu-intel commented:
> > Hi, a workaround as follows can be implemented, which will resolve the `Unsupported level` exception. This is just adding Level 4 as a placeholder for any kind of `Unsupported Datatype Optimizations` in the training graph transformer utils.
> 
> It's good to just add a placeholder for training part.
> 
> For DNNL, maybe use `#ifdef USE_DNNL` and force to skip level 4 and loop for DNNL.

Hi,

I will soon be pushing the changes we discussed. As of now, we are testing the changes with the systems available at our end.

#### 1. With the new changes, I added a corner case unittest where a Cast node output is also a graph output. Thanks for the suggestion; with my assumptions of counting output edges, I missed that check.

#### 2. I also added Level-4 placeholders in graph transforms utils under orttraining. This will help to resolve the following exception.

```
C++ exception with description "/onnxruntime_src/orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:220 Unsupported level 4
```

#### With the last update, the following builds are failing.

``` 
Linux CI / Build Linux x64 Release with training / build_test_pipeline (pull_request)
Linux DNNL CI / Build Linux x64 Release (DNNL EP) / build_test_pipeline (pull_request)
```

#### With the latest changes, only the errors related to the Linux DNNL CI build are yet to be resolved, as I am not able to reproduce the errors on my side. The following are the steps I took to reproduce these errors with my system.

##### A) Build Command

```
./build.sh --config Release --cmake_generator Ninja --skip_submodule_sync --build_shared_lib --parallel --enable_onnx_tests --use_dnnl --skip_test
```
> NOTE: With flags like  --use_vcpkg, --use_vcpkg_ms_internal_asset_cache, --use_binskim_compliant_compile_flags, --build_wheel, --build_nuget, and --test, we are facing errors and unable to build successfully on a Linux x64 system.

##### B) Test Result

> Some tests are failing, and none of them seem to be related to this PR, as they are also failing with main branch. 

```
[----------] Global test environment tear-down
[==========] 4955 tests from 315 test suites ran. (939362 ms total)
[  PASSED  ] 4920 tests.
[  SKIPPED ] 2 tests, listed below:
[  SKIPPED ] MatMulFpQ4.MatMul2DSym
[  SKIPPED ] MatMulFpQ4.MatMul2DBlkZp
[  FAILED  ] 33 tests, listed below:
[  FAILED  ] ONNXModelsTest.squeeze_net
[  FAILED  ] ONNXModelsTest.bvlc_alexnet_1
[  FAILED  ] InferenceSessionTests.ModelMetadata
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/0, where GetParam() = "bvlc_alexnet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/1, where GetParam() = "bvlc_googlenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/2, where GetParam() = "bvlc_reference_caffenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/3, where GetParam() = "bvlc_reference_rcnn_ilsvrc13"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/4, where GetParam() = "densenet121"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/5, where GetParam() = "emotion_ferplus"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/6, where GetParam() = "inception_v1"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/7, where GetParam() = "inception_v2"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/8, where GetParam() = "mnist"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/9, where GetParam() = "resnet50"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/10, where GetParam() = "shufflenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/11, where GetParam() = "squeezenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/12, where GetParam() = "tiny_yolov2"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/13, where GetParam() = "vgg19"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromFile/14, where GetParam() = "zfnet512"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/0, where GetParam() = "bvlc_alexnet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/1, where GetParam() = "bvlc_googlenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/2, where GetParam() = "bvlc_reference_caffenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/3, where GetParam() = "bvlc_reference_rcnn_ilsvrc13"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/4, where GetParam() = "densenet121"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/5, where GetParam() = "emotion_ferplus"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/6, where GetParam() = "inception_v1"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/7, where GetParam() = "inception_v2"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/8, where GetParam() = "mnist"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/9, where GetParam() = "resnet50"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/10, where GetParam() = "shufflenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/11, where GetParam() = "squeezenet"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/12, where GetParam() = "tiny_yolov2"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/13, where GetParam() = "vgg19"
[  FAILED  ] ONNXModelsTests/ONNXModelsTest1.LoadFromProtobuf/14, where GetParam() = "zfnet512"

33 FAILED TESTS
  YOU HAVE 6 DISABLED TESTS
```

##### C) Individual test results

> These tests, which are failing with the Linux DNNL CI build in the workflow checks, are not failing on our side.

```
$ ./onnxruntime_test_all --gtest_filter="ActivationOpTest.LeakyRelu_bfloat16"
Note: Google Test filter = ActivationOpTest.LeakyRelu_bfloat16
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ActivationOpTest
[ RUN      ] ActivationOpTest.LeakyRelu_bfloat16
2025-05-29 15:15:42.056000335 [W:onnxruntime:Default, activation_op_test.cc:362 TestBody] Hardware does NOT support BF16
[       OK ] ActivationOpTest.LeakyRelu_bfloat16 (1 ms)
[----------] 1 test from ActivationOpTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 1 test.

$ ./onnxruntime_test_all --gtest_filter="QAttentionTest.QAttentionDNNLBatch1"
Note: Google Test filter = QAttentionTest.QAttentionDNNLBatch1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QAttentionTest
[ RUN      ] QAttentionTest.QAttentionDNNLBatch1
[       OK ] QAttentionTest.QAttentionDNNLBatch1 (50 ms)
[----------] 1 test from QAttentionTest (50 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (50 ms total)
[  PASSED  ] 1 test.

$ ./onnxruntime_test_all --gtest_filter="QAttentionTest.QAttentionDNNLBatch2"
Note: Google Test filter = QAttentionTest.QAttentionDNNLBatch2
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QAttentionTest
[ RUN      ] QAttentionTest.QAttentionDNNLBatch2
[       OK ] QAttentionTest.QAttentionDNNLBatch2 (50 ms)
[----------] 1 test from QAttentionTest (50 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (50 ms total)
[  PASSED  ] 1 test.

$ ./onnxruntime_test_all --gtest_filter="QAttentionTest.QAttentionDNNLMaskPartialSequence"
Note: Google Test filter = QAttentionTest.QAttentionDNNLMaskPartialSequence
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QAttentionTest
[ RUN      ] QAttentionTest.QAttentionDNNLMaskPartialSequence
[       OK ] QAttentionTest.QAttentionDNNLMaskPartialSequence (50 ms)
[----------] 1 test from QAttentionTest (50 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (50 ms total)
[  PASSED  ] 1 test.

$ ./onnxruntime_test_all --gtest_filter="QAttentionTest.QAttentionNoMaskIndex"
Note: Google Test filter = QAttentionTest.QAttentionNoMaskIndex
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QAttentionTest
[ RUN      ] QAttentionTest.QAttentionNoMaskIndex
[       OK ] QAttentionTest.QAttentionNoMaskIndex (51 ms)
[----------] 1 test from QAttentionTest (51 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (51 ms total)
[  PASSED  ] 1 test.

$ ./onnxruntime_test_all --gtest_filter="QAttentionTest.QAttentionPrunedModel"
Note: Google Test filter = QAttentionTest.QAttentionPrunedModel
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from QAttentionTest
[ RUN      ] QAttentionTest.QAttentionPrunedModel
[       OK ] QAttentionTest.QAttentionPrunedModel (56 ms)
[----------] 1 test from QAttentionTest (56 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (56 ms total)
[  PASSED  ] 1 test.
```

#### I am planning to upload all the changes we have done so far, and then check if it causes the same issue with the Linux DNNL CI build. Let me know your thoughts on the same.

tianleiwu commented:
Sounds good to me. Please merge the latest main branch, which has some fix for CI pipeline.

sunnyshu-intel commented:
> Sounds good to me. Please merge the latest main branch, which has some fix for CI pipeline.

Hi, I uploaded the latest changes. Once the workflow checks are re-run, I will look into any remaining build errors asap. Thanks.

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline

sunnyshu-intel commented:
> /azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline

#### Hi, it seems Linux DNNL CI build errors are the only ones remaining to be solved.  I have tried looking into it once more, however it doesn't seem to be caused by this PR. 

![image](https://github.com/user-attachments/assets/429fc4fa-82e8-4efd-9b62-d4a2b5da0104)

- I will upload a few changes to resolve the `cpplint` issues and a `misspell` issue as reported by Github Actions in the `Files changed` tab with an updated commit.

- I am also adding a logging change as follows for a cleaner way to trace the graph optimization loop operations.

```diff
diff --git a/onnxruntime/core/session/inference_session.cc b/onnxruntime/core/session/inference_session.cc
index 8ab9bdcf5f..4cff20c488 100644
--- a/onnxruntime/core/session/inference_session.cc
+++ b/onnxruntime/core/session/inference_session.cc
@@ -1423,8 +1423,11 @@ common::Status InferenceSession::TransformGraph(onnxruntime::Graph& graph, bool
   for (int steps = 0; steps < 10; steps++) {
     // Warning that we are running this loop too many times
     if (steps >= 3) {
-      LOGS(*session_logger_, WARNING) << "Running graph optimizations in loop " << (steps + 1) << " times."
+      LOGS(*session_logger_, WARNING) << "Running graph optimizations in loop " << steps << " time/s"
                                       << " (Graph Optimizations Loop Level : " << graph_optimizations_loop_level << ")";
+    } else {
+      LOGS(*session_logger_, INFO) << "Running graph optimizations in loop " << steps << " time/s"
+                                   << " (Graph Optimizations Loop Level : " << graph_optimizations_loop_level << ")";
     }
 
```

#### However, before all of that, I would like to focus on Linux DNNL CI build errors. For the same,  I would like to hear your thoughts on how to proceed. Thanks.

tianleiwu commented:
/azp run Linux QNN CI Pipeline, Win_TRT_Minimal_CUDA_Test_CI, Windows ARM64 QNN CI Pipeline, Windows GPU Doc Gen CI Pipeline, Windows x64 QNN CI Pipeline
